File: php/testcode/xxe_004.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_domdocument_safe
function xxe004(BenchmarkRequest $req): BenchmarkResponse {
    $xml = $req->bodyStr();
    $doc = new DOMDocument();
    $doc->loadXML($xml, LIBXML_NOENT | LIBXML_NONET); // vuln-code-snippet safe-line php_xxe_domdocument_safe
    $name = $doc->getElementsByTagName('name')->item(0)->textContent ?? '';
    return BenchmarkResponse::ok($name);
}
// vuln-code-snippet end php_xxe_domdocument_safe

File: ruby/testcode/xxe_004.rb
require_relative 'shared'

module REXML
  class Document
    def initialize(input)
      @input = input
    end
    def to_s; @input.to_s; end
  end
  module Security
    def self.entity_expansion_limit=(val); end
  end
end

# vuln-code-snippet start ruby_xxe_rexml_limit
def parse_rexml_safe(req)
  xml_input = req.body_str
  REXML::Security.entity_expansion_limit = 0 # vuln-code-snippet safe-line ruby_xxe_rexml_limit
  doc = REXML::Document.new(xml_input)
  BenchmarkResponse.ok(doc.to_s)
end
# vuln-code-snippet end ruby_xxe_rexml_limit
